Automattic/woocommerce-payments
# Update cache after persisting the User session

Fixes https://github.com/Automattic/woopay/issues/2774 

#### Changes proposed in this Pull Request

When a user clicks the “Buy with WooPay” button from the product page, the product is added to the cart, and an empty WC session is created. The session identifier is sent to WooPay and used for Store API requests. These requests are handled by the `WCPay\Platform_Checkout\SessionHandler` class although when visiting the checkout page, the session is handled by the `WC_Session_Handler` class.

As a result, we end up with two classes managing the same WC session.

`WC_Session_Handler` tries to retrieve the session from the cache after it has been updated by WooPay via the Store API. However, it fetches an outdated state of the session (last updated when the user clicked the “Buy with WooPay” button), causing the checkout pages to appear empty.

To resolve this, we replicate the steps performed by the `WC_Session_Handler` in its `save_data` method. We update the cache in `WCPay\Platform_Checkout\SessionHandler` via `wp_cache_set`, ensuring that the session is synchronized and ready for the consumption of the `WC_Session_Handler` class.









### Testing instructions


> [!IMPORTANT]  
> In order to get the bug we need to enable object cache. An easy way to do that is by installing the [W3 Total Cache plugin](https://pe.wordpress.org/plugins/w3-total-cache/). Go through its settings and make sure to enable the Object Cache with the save to "Disk" option.

**Required setup**
- WooPay is enabled.
- Make sure your site W3 Total Cache plugin is properly set up with Object Cache enabled.
- Make sure you are logged into your local WooPay account.
- Make sure you have both, block-based and a shortcode-based checkout pages.
- Make sure you have shipping options defined.

**Reproduce the bug.**

1. Checkout to the `develop` branch.
2. Go to the merchant store and click on a simple product that requires shipping.
3. Click "Buy with WooPay"
4. In WooPay expand the "Pay with" section. Then click "Alternative payment methods"
5. Notice your WooPay customer details are not pre-filled in the checkout form.
6. Go to the other checkout page and notice customer details are also missing.

**Bug fix confirmation.**
1. Checkout to the `as-fix-woopay-data-mismatch` branch.
2. Remove cookies in your merchant store to ensure you are creating a new WC session.
2. Go to the store and click on a simple product that requires shipping.
3. Click "Buy with WooPay"
4. In WooPay expand the "Pay with" section. Then click "Alternative payment methods"
5. Notice your WooPay customer details **are pre-filled** in the checkout form.
6. Go to the other checkout page and notice customer details are pe-filled correctly.

#### Note
- The billing address is not currently populated on the block-based checkout page. It used to work, but it seems that recent changes to the billing area have made it no longer work. It's behaving like this even without Object Cache, so it's a separate issue. Shortcode-based checkout should work fine.

-------------------


**Post merge**
